src/client/play.js

```diff
@@ -5,15 +5,37 @@ const concat = require('../transforms/binaryStream').concat
 module.exports = function (client, options) {
   client.once('success', onLogin)
 
+  const mcData = require('minecraft-data')(client.version)
+
   function onLogin (packet) {
     client.state = states.PLAY
     client.uuid = packet.uuid
     client.username = packet.username
-    client.signMessage = (message, timestamp, salt = 0) => {
+    client.signMessage = (message, timestamp, salt = 0n, isCommand, preview) => {
       if (!client.profileKeys) throw Error("Can't sign message without profile keys, please set valid auth mode")
-      const signable = concat('i64', salt, 'UUID', client.uuid, 'i64',
-        timestamp / 1000n, 'pstring', JSON.stringify({ text: message }))
-      return crypto.sign('RSA-SHA256', signable, client.profileKeys.private)
+      if (mcData.supportFeature('chainedSignature')) { // 1.19.1/1.19.2
+        const hashable = crypto.createHash('sha256').update(concat('i64', salt, 'i64', timestamp / 1000n, 'pstring', message, 'i8', 70))
+        if (preview) hashable.update(Buffer.from(preview, 'utf8'))
+        const hash = hashable.digest()
+        const proto = ['UUID', client.uuid, 'buffer', hash]
+        if (!!client.lastSignature && !isCommand) proto.unshift('buffer', client.lastSignature)
+        const signable = concat(...proto)
+        const signed = crypto.sign('RSA-SHA256', signable, client.profileKeys.private)
+        if (!isCommand) client.lastSignature = signed
+        return signed
+      } else if (mcData.supportFeature('sessionSignature')) { // 1.19.3
+        if (!client.sessionUUID) throw Error("Can't sign message before initializing chat session")
+        if (!client.sessionIndex) client.sessionIndex = 0
+        const length = Buffer.byteLength(message, 'utf8')
+        const signable = concat('i32', 1, 'UUID', client.uuid, 'UUID', client.sessionUUID, 'i32', client.sessionIndex, 'i64', salt, 'i64', timestamp / 1000n, 'i32', length, 'pstring', message, 'i32', 0)
```

Comment:
explain the 1 and 0 at the end?

Reply:
Same thing as with the 70 in proto 760. For some reason those values are padded in before signing (and before verifying). They must be present or the client will get kicked for invalid or missing signature

Reply:
Gonna have to add: Trailing 0 is actually the length of the last seen messages array. client.signMessage currently does not account for chaining at all (as the server doesn't seem to mind that the client does not send any acknowledgements), although it should definitely be implemented to mimic the vanilla client